client/http/client.go

```diff
@@ -140,42 +141,55 @@ func (ci *clientInner) execDuration(name string, duration time.Duration) {
 
 // requestWithRetry will first try to send the request to the PD leader, if it fails, it will try to send
 // the request to the other PD followers to gain a better availability.
-// TODO: support custom retry logic, e.g. retry with customizable backoffer.
 func (ci *clientInner) requestWithRetry(
 	ctx context.Context,
 	reqInfo *requestInfo,
 	headerOpts ...HeaderOption,
 ) error {
 	var (
-		statusCode             int
-		err                    error
-		addr                   string
-		pdAddrs, leaderAddrIdx = ci.getPDAddrs()
+		statusCode int
+		err        error
 	)
-	// Try to send the request to the PD leader first.
-	if leaderAddrIdx != -1 {
-		addr = pdAddrs[leaderAddrIdx]
-		statusCode, err = ci.doRequest(ctx, addr, reqInfo, headerOpts...)
-		if err == nil || noNeedRetry(statusCode) {
-			return err
+	execFunc := func() error {
+		var (
+			addr                   string
+			pdAddrs, leaderAddrIdx = ci.getPDAddrs()
+		)
+		// Try to send the request to the PD leader first.
+		if leaderAddrIdx != -1 {
+			addr = pdAddrs[leaderAddrIdx]
+			statusCode, err = ci.doRequest(ctx, addr, reqInfo, headerOpts...)
+			if err == nil || noNeedRetry(statusCode) {
+				return err
+			}
+			log.Debug("[pd] request leader addr failed",
+				zap.String("source", ci.source), zap.Int("leader-idx", leaderAddrIdx), zap.String("addr", addr), zap.Error(err))
 		}
-		log.Debug("[pd] request leader addr failed",
-			zap.String("source", ci.source), zap.Int("leader-idx", leaderAddrIdx), zap.String("addr", addr), zap.Error(err))
-	}
-	// Try to send the request to the other PD followers.
-	for idx := 0; idx < len(pdAddrs); idx++ {
-		if idx == leaderAddrIdx {
-			continue
+		// Try to send the request to the other PD followers.
+		for idx := 0; idx < len(pdAddrs); idx++ {
+			if idx == leaderAddrIdx {
+				continue
+			}
+			addr = pdAddrs[idx]
```

Comment:
The crucial line change is from "addr = ci.pdAddrs[idx]" to "addr = pdAddrs[idx]".